ast/walk.go

```diff
@@ -72,6 +72,112 @@ func Walk(expr Expression, visitor func(Expression) bool) bool {
 	panic(fmt.Sprintf("Unexpected AST expression of type %T: %[1]v", expr))
 }
 
+// WalkAndFindParentUsageInvalid is like Walk() but it returns an error if an invalid
+// parent usage is found.
+// This function is exclusively used to walk the body of a custom defined function.
+// The usage of a parent operator is *not* allowed in the body of a custom defined function
+// if it reaches out of that function's scope. It's valid within a filter or a projection, because
+// we create NewNestedScope for those.
+func WalkAndFindParentUsageInvalid(count int, expr Expression) (int, error) {
+	if expr == nil {
+		return count, nil
+	}
+	switch e := expr.(type) {
+	case *Parent:
+		count--
+		if count <= 0 {
+			return count, fmt.Errorf("Parent usage is invalid")
+		}
+		return count, nil
```

Comment:
Shouldn't the check for `count` come _before_ the increment? 🤔 

```suggestion
		if count <= 0 {
			return count, fmt.Errorf("Parent usage is invalid")
		}
		return count-1, nil
```